src/fast_array_utils/_plugins/numba_sparse.py

```diff
@@ -0,0 +1,237 @@
+
+
+def make_typeof_fn(typ: type[CS2DType]) -> Callable[[CSBase, _TypeofContext], CS2DType]:
+    def typeof(val: CSBase, c: _TypeofContext) -> CS2DType:
+        data = cast("nbtypes.Array", typeof_impl(val.data, c))
+        return typ(data.dtype)
+
+    return typeof
+
+
+if TYPE_CHECKING:
+    _Base = models.StructModel[CS2DType]
+else:
+    _Base = models.StructModel
+
+
+class CS2DModel(_Base):
+    def __init__(self, dmm: object, fe_type: CS2DType) -> None:
+        members = [
+            ("data", fe_type.data),
+            ("indices", fe_type.indices),
+            ("indptr", fe_type.indptr),
+            ("shape", fe_type.shape),
+        ]
+        super().__init__(dmm, fe_type, members)
+
+
+CLASSES = [sparse.csr_matrix, sparse.csc_matrix, sparse.csr_array, sparse.csc_array]
+TYPES: list[type[CS2DType]] = [
+    type(f"{cls.__name__}Type", (CS2DType,), {"cls": cls, "name": cls.__name__}) for cls in CLASSES
+]
+TYPEOF_FUNCS = {typ.cls: make_typeof_fn(typ) for typ in TYPES}
+MODELS = {typ: type(f"{typ.cls.__name__}Model", (CS2DModel,), {}) for typ in TYPES}
+
+
+def unbox_matrix(typ: CS2DType, obj: Value, c: UnboxContext) -> NativeValue:
+    struct_proxy_cls = cgutils.create_struct_proxy(typ)
+    struct_ptr = struct_proxy_cls(c.context, c.builder)
+
+    data = c.pyapi.object_getattr_string(obj, "data")
+    indices = c.pyapi.object_getattr_string(obj, "indices")
+    indptr = c.pyapi.object_getattr_string(obj, "indptr")
+    shape = c.pyapi.object_getattr_string(obj, "shape")
+
+    struct_ptr.data = c.unbox(typ.data, data).value
+    struct_ptr.indices = c.unbox(typ.indices, indices).value
+    struct_ptr.indptr = c.unbox(typ.indptr, indptr).value
+    struct_ptr.shape = c.unbox(typ.shape, shape).value
+
+    c.pyapi.decref(data)
+    c.pyapi.decref(indices)
+    c.pyapi.decref(indptr)
+    c.pyapi.decref(shape)
+
+    is_error_ptr = cgutils.alloca_once_value(c.builder, cgutils.false_bit)
+    is_error = c.builder.load(is_error_ptr)
+
+    return NativeValue(struct_ptr._getvalue(), is_error=is_error)  # noqa: SLF001
+
+
+def box_matrix(typ: CS2DType, val: NativeValue, c: BoxContext) -> Value:
+    struct_proxy_cls = cgutils.create_struct_proxy(typ)
+    struct_ptr = struct_proxy_cls(c.context, c.builder, value=val)
+
+    data_obj = c.box(typ.data, struct_ptr.data)
+    indices_obj = c.box(typ.indices, struct_ptr.indices)
+    indptr_obj = c.box(typ.indptr, struct_ptr.indptr)
+    shape_obj = c.box(typ.shape, struct_ptr.shape)
+
+    c.pyapi.incref(data_obj)
+    c.pyapi.incref(indices_obj)
+    c.pyapi.incref(indptr_obj)
+    c.pyapi.incref(shape_obj)
+
+    cls_obj = c.pyapi.unserialize(c.pyapi.serialize_object(typ.instance_class))
+    obj = c.pyapi.call_function_objargs(cls_obj, (data_obj, indices_obj, indptr_obj, shape_obj))
+
+    c.pyapi.decref(data_obj)
+    c.pyapi.decref(indices_obj)
+    c.pyapi.decref(indptr_obj)
+    c.pyapi.decref(shape_obj)
+
+    return obj
+
+
+@overload(np.shape)
```

Comment:
Could you explain this a bit? I understand what the point of this overload is I think (giving `CS2DType` a `shape` implementation`) but I don't understand why doing it this way accomplishes that